
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>XICRA.scripts.RNAbiotype &#8212; XICRA v0.9.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/alabaster.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css" />
    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
   
  <link rel="stylesheet" href="../../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for XICRA.scripts.RNAbiotype</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1">############################################################</span>
<span class="c1">## Jose F. Sanchez                                        ##</span>
<span class="c1">## Copyright (C) 2019-2020 Lauro Sumoy Lab, IGTP, Spain   ##</span>
<span class="c1">############################################################</span>
<span class="c1">## useful imports</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">io</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">sys</span> <span class="kn">import</span> <span class="n">argv</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">from</span> <span class="nn">termcolor</span> <span class="kn">import</span> <span class="n">colored</span>
<span class="kn">import</span> <span class="nn">concurrent.futures</span>

<span class="c1">## import my modules</span>
<span class="kn">from</span> <span class="nn">XICRA.config</span> <span class="kn">import</span> <span class="n">set_config</span>

<span class="c1">## import HCGB</span>
<span class="kn">from</span> <span class="nn">HCGB.functions</span> <span class="kn">import</span> <span class="n">system_call_functions</span><span class="p">,</span> <span class="n">main_functions</span><span class="p">,</span> <span class="n">time_functions</span>
<span class="kn">from</span> <span class="nn">HCGB.functions</span> <span class="kn">import</span> <span class="n">files_functions</span><span class="p">,</span> <span class="n">math_functions</span>

<span class="c1">## plots</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>

<span class="n">matplotlib</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s1">&#39;agg&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">pandas.plotting</span> <span class="kn">import</span> <span class="n">table</span>

<span class="c1">#####################</span>
<div class="viewcode-block" id="help_info"><a class="viewcode-back" href="../../../devel/scripts/RNAbiotype.html#XICRA.scripts.RNAbiotype.help_info">[docs]</a><span class="k">def</span> <span class="nf">help_info</span><span class="p">():</span>
	<span class="sd">&#39;&#39;&#39;Provide information on RNA biotype analysis&#39;&#39;&#39;</span>
	<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;** TODO: Add description on RNAbiotype analysis and details&quot;</span><span class="p">)</span>
	<span class="n">exit</span><span class="p">()</span></div>

<span class="c1">#####################</span>
<div class="viewcode-block" id="biotype_all"><a class="viewcode-back" href="../../../devel/scripts/RNAbiotype.html#XICRA.scripts.RNAbiotype.biotype_all">[docs]</a><span class="k">def</span> <span class="nf">biotype_all</span><span class="p">(</span><span class="n">featureCount_exe</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">threads</span><span class="p">,</span> <span class="n">Debug</span><span class="p">,</span> <span class="n">allow_multimap</span><span class="p">,</span> <span class="n">stranded</span><span class="p">):</span>
	
	<span class="c1">## folder for results</span>
	<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">path</span><span class="p">):</span>
		<span class="n">files_functions</span><span class="o">.</span><span class="n">create_folder</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>

	<span class="n">out_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="s1">&#39;featureCount.out&#39;</span><span class="p">)</span>
	<span class="n">logfile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">name</span> <span class="o">+</span> <span class="s1">&#39;_RNAbiotype.log&#39;</span><span class="p">)</span>

	<span class="n">filename_stamp_all</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;/.success_all&#39;</span>
	<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename_stamp_all</span><span class="p">):</span>
		<span class="n">stamp</span> <span class="o">=</span> <span class="n">time_functions</span><span class="o">.</span><span class="n">read_time_stamp</span><span class="p">(</span><span class="n">filename_stamp_all</span><span class="p">)</span>
		<span class="nb">print</span> <span class="p">(</span><span class="n">colored</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">A previous command generated results on: </span><span class="si">%s</span><span class="s2"> [</span><span class="si">%s</span><span class="s2"> -- </span><span class="si">%s</span><span class="s2">]&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">stamp</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s1">&#39;RNAbiotype&#39;</span><span class="p">),</span> <span class="s1">&#39;yellow&#39;</span><span class="p">))</span>
		<span class="k">return</span><span class="p">()</span>

	<span class="k">else</span><span class="p">:</span>
		<span class="n">filename_stamp_featureCounts</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;/.success_featureCounts&#39;</span>
		<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename_stamp_featureCounts</span><span class="p">):</span>
			<span class="n">stamp</span> <span class="o">=</span> <span class="n">time_functions</span><span class="o">.</span><span class="n">read_time_stamp</span><span class="p">(</span><span class="n">filename_stamp_featureCounts</span><span class="p">)</span>
			<span class="nb">print</span> <span class="p">(</span><span class="n">colored</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">A previous command generated results on: </span><span class="si">%s</span><span class="s2"> [</span><span class="si">%s</span><span class="s2"> -- </span><span class="si">%s</span><span class="s2">]&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">stamp</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s1">&#39;featureCounts&#39;</span><span class="p">),</span> <span class="s1">&#39;yellow&#39;</span><span class="p">))</span>
		<span class="k">else</span><span class="p">:</span>

			<span class="c1">## debugging messages</span>
			<span class="k">if</span> <span class="n">Debug</span><span class="p">:</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;** DEBUG:&quot;</span><span class="p">)</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;featureCounts system call for sample: &quot;</span> <span class="o">+</span> <span class="n">name</span><span class="p">)</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;out_file: &quot;</span> <span class="o">+</span> <span class="n">out_file</span><span class="p">)</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;logfile: &quot;</span> <span class="o">+</span> <span class="n">logfile</span><span class="p">)</span>
		
			<span class="c1">## send command for feature count</span>
			<span class="c1">## Allow multimapping</span>
			<span class="k">if</span> <span class="n">allow_multimap</span><span class="p">:</span>
				<span class="n">cmd_featureCount</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> -s </span><span class="si">%s</span><span class="s1"> -M -O -T </span><span class="si">%s</span><span class="s1"> -p -t exon -g transcript_biotype -a </span><span class="si">%s</span><span class="s1"> -o </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> 2&gt; </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span><span class="p">(</span>
					<span class="n">featureCount_exe</span><span class="p">,</span> <span class="n">stranded</span><span class="p">,</span> <span class="n">threads</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">,</span> <span class="n">out_file</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span> <span class="n">logfile</span><span class="p">)</span>
				<span class="p">)</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">cmd_featureCount</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> -s </span><span class="si">%s</span><span class="s1"> --largestOverlap -T </span><span class="si">%s</span><span class="s1"> -p -t exon -g transcript_biotype -a </span><span class="si">%s</span><span class="s1"> -o </span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1"> 2&gt; </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span><span class="p">(</span>
					<span class="n">featureCount_exe</span><span class="p">,</span> <span class="n">stranded</span><span class="p">,</span> <span class="n">threads</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">,</span> <span class="n">out_file</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span> <span class="n">logfile</span><span class="p">)</span>
				<span class="p">)</span>
				
				
			<span class="c1">## system call</span>
			<span class="n">cmd_featureCount_code</span> <span class="o">=</span> <span class="n">system_call_functions</span><span class="o">.</span><span class="n">system_call</span><span class="p">(</span><span class="n">cmd_featureCount</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
			<span class="k">if</span> <span class="ow">not</span> <span class="n">cmd_featureCount_code</span><span class="p">:</span>
				<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;** ERROR: featureCount failed for sample &quot;</span> <span class="o">+</span> <span class="n">name</span><span class="p">)</span>
				<span class="n">exit</span><span class="p">()</span>
				
			<span class="c1">## print time stamp</span>
			<span class="n">time_functions</span><span class="o">.</span><span class="n">print_time_stamp</span><span class="p">(</span><span class="n">filename_stamp_featureCounts</span><span class="p">)</span>
		
		<span class="c1">## parse results</span>
		<span class="p">(</span><span class="n">extended_Stats_file</span><span class="p">,</span> <span class="n">RNAbiotypes_stats_file</span><span class="p">)</span> <span class="o">=</span> <span class="n">parse_featureCount</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span> <span class="n">Debug</span><span class="p">)</span>
		
		<span class="c1">## debugging messages</span>
		<span class="k">if</span> <span class="n">Debug</span><span class="p">:</span>
			<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;** DEBUG:&quot;</span><span class="p">)</span>
			<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;extended_Stats: &quot;</span> <span class="o">+</span> <span class="n">extended_Stats_file</span><span class="p">)</span>
			<span class="nb">print</span> <span class="p">(</span><span class="n">main_functions</span><span class="o">.</span><span class="n">get_data</span><span class="p">(</span><span class="n">extended_Stats_file</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;header=None&#39;</span><span class="p">))</span>
			<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;RNAbiotypes_stats: &quot;</span> <span class="o">+</span> <span class="n">RNAbiotypes_stats_file</span><span class="p">)</span>
			<span class="nb">print</span> <span class="p">(</span><span class="n">main_functions</span><span class="o">.</span><span class="n">get_data</span><span class="p">(</span><span class="n">RNAbiotypes_stats_file</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;header=None&#39;</span><span class="p">))</span>

	<span class="k">return</span> <span class="p">()</span></div>

<span class="c1">#######################################################################</span>
<div class="viewcode-block" id="parse_featureCount"><a class="viewcode-back" href="../../../devel/scripts/RNAbiotype.html#XICRA.scripts.RNAbiotype.parse_featureCount">[docs]</a><span class="k">def</span> <span class="nf">parse_featureCount</span><span class="p">(</span><span class="n">out_file</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span> <span class="n">Debug</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Parses featureCount results for RNAbiotype analysis.</span>
<span class="sd">	</span>
<span class="sd">	:param out_file: Name provided to featureCount for output results.</span>
<span class="sd">	:param path:</span>
<span class="sd">	:param name:</span>
<span class="sd">	</span>
<span class="sd">	</span>
<span class="sd">	&quot;&quot;&quot;</span>

	<span class="c1">## file names</span>
	<span class="n">out_tsv_file_name</span> <span class="o">=</span> <span class="n">out_file</span> <span class="o">+</span> <span class="s1">&#39;.tsv&#39;</span>
	<span class="n">RNA_biotypes_file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">name</span> <span class="o">+</span> <span class="s1">&#39;_RNAbiotype.tsv&#39;</span><span class="p">)</span>

	<span class="c1">##</span>
	<span class="n">filename_stamp_parse</span> <span class="o">=</span> <span class="n">path</span> <span class="o">+</span> <span class="s1">&#39;/.success_parse&#39;</span>
	<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename_stamp_parse</span><span class="p">):</span>
		<span class="n">stamp</span> <span class="o">=</span> <span class="n">time_functions</span><span class="o">.</span><span class="n">read_time_stamp</span><span class="p">(</span><span class="n">filename_stamp_parse</span><span class="p">)</span>
		<span class="nb">print</span> <span class="p">(</span><span class="n">colored</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">A previous command generated results on: </span><span class="si">%s</span><span class="s2"> [</span><span class="si">%s</span><span class="s2"> -- </span><span class="si">%s</span><span class="s2">]&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">stamp</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s1">&#39;parse results&#39;</span><span class="p">),</span> <span class="s1">&#39;yellow&#39;</span><span class="p">))</span>
	<span class="k">else</span><span class="p">:</span>
	
		<span class="c1">## debugging messages</span>
		<span class="k">if</span> <span class="n">Debug</span><span class="p">:</span>
			<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;** DEBUG:&quot;</span><span class="p">)</span>
			<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;Parse results for sample: &quot;</span> <span class="o">+</span> <span class="n">name</span><span class="p">)</span>
			
		<span class="c1">## parse results</span>
		<span class="n">out_tsv_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_tsv_file_name</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
		<span class="n">RNA_biotypes_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">RNA_biotypes_file_name</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
		<span class="n">tRNA_count</span> <span class="o">=</span> <span class="mi">0</span>
		
		<span class="c1">##########################################</span>
		<span class="c1">### read count file</span>
		<span class="c1">##########################################</span>
		<span class="n">count_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_file</span><span class="p">)</span>
		<span class="n">count_file_text</span> <span class="o">=</span> <span class="n">count_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
		<span class="n">count_file_lines</span> <span class="o">=</span> <span class="n">count_file_text</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>	
	
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">count_file_lines</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span>
				<span class="k">continue</span>
			<span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;Geneid&#39;</span><span class="p">):</span>
				<span class="k">continue</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="n">ID</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
				<span class="n">count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
				<span class="n">string2write_raw</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">ID</span><span class="p">,</span> <span class="n">count</span><span class="p">)</span>
				<span class="n">out_tsv_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">string2write_raw</span><span class="p">)</span>
	
				<span class="n">tRNA_search</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;.*tRNA&quot;</span><span class="p">,</span> <span class="n">ID</span><span class="p">)</span>
				<span class="k">if</span> <span class="n">tRNA_search</span><span class="p">:</span>
					<span class="n">tRNA_count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">tRNA_count</span><span class="p">)</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>				
				<span class="k">elif</span> <span class="p">(</span><span class="n">count</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
					<span class="n">RNA_biotypes_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">string2write_raw</span><span class="p">)</span>
		
		<span class="c1">## count and summary tRNA</span>
		<span class="n">string2write</span> <span class="o">=</span> <span class="s2">&quot;tRNA</span><span class="se">\t</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span><span class="n">tRNA_count</span>
		<span class="n">RNA_biotypes_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">string2write</span><span class="p">)</span>
		<span class="n">RNA_biotypes_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
				
		<span class="c1">##########################################</span>
		<span class="c1">### read summary count file</span>
		<span class="c1">##########################################</span>
		<span class="n">summary_count_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">out_file</span> <span class="o">+</span> <span class="s1">&#39;.summary&#39;</span><span class="p">)</span>
		<span class="n">summary_count_file_text</span> <span class="o">=</span> <span class="n">summary_count_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
		<span class="n">summary_count_file_lines</span> <span class="o">=</span> <span class="n">summary_count_file_text</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>	
	
		<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">summary_count_file_lines</span><span class="p">:</span>
			<span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;Status&#39;</span><span class="p">):</span>
				<span class="k">continue</span>
			<span class="k">elif</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;Assigned&#39;</span><span class="p">):</span>
				<span class="k">continue</span>
			<span class="k">else</span><span class="p">:</span>
				<span class="c1">## adds Unassigned_Ambiguity</span>
				<span class="c1">## adds Unassigned_NoFeatures</span>
				<span class="n">ID</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
				<span class="n">count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
	
				<span class="c1">## skip empty entries</span>
				<span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
					<span class="k">continue</span>
				<span class="n">string2write_raw</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">%s</span><span class="se">\t</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">ID</span><span class="p">,</span> <span class="n">count</span><span class="p">)</span>
				<span class="n">out_tsv_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">string2write_raw</span><span class="p">)</span>
	
		<span class="c1">##########################################</span>
		<span class="c1">## get mapping statistics according to mapping software</span>
		<span class="c1">##########################################</span>
		<span class="n">count_multi</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="n">count_unmap</span> <span class="o">=</span> <span class="mi">0</span>
		<span class="n">mapping_folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">bam_file</span><span class="p">)</span>
		<span class="n">mapping_stats</span> <span class="o">=</span> <span class="n">mapping_folder</span> <span class="o">+</span> <span class="s1">&#39;/Log.final.out&#39;</span>
		
		<span class="c1">## -------------------------------- ##</span>
		<span class="c1">### STAR mapping		</span>
		<span class="c1">## -------------------------------- ##</span>
		<span class="k">if</span> <span class="n">files_functions</span><span class="o">.</span><span class="n">is_non_zero_file</span><span class="p">(</span><span class="n">mapping_stats</span><span class="p">):</span>
			<span class="c1">## debugging messages</span>
			<span class="k">if</span> <span class="n">Debug</span><span class="p">:</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;** DEBUG:&quot;</span><span class="p">)</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;STAR mapping available for sample: &quot;</span> <span class="o">+</span> <span class="n">name</span><span class="p">)</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;mapping_folder: &quot;</span> <span class="o">+</span> <span class="n">mapping_folder</span><span class="p">)</span>
	
			<span class="n">mapping_stats_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">mapping_stats</span><span class="p">)</span>
			<span class="n">mapping_stats_file_text</span> <span class="o">=</span> <span class="n">mapping_stats_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
			<span class="n">mapping_stats_file_lines</span> <span class="o">=</span> <span class="n">mapping_stats_file_text</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>	
	
			<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">mapping_stats_file_lines</span><span class="p">:</span>
				<span class="n">multi_search</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;.*Number of reads mapped to&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
				<span class="n">unmap_search</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;.*unmapped.*&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
				<span class="n">input_search</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;.*input reads.*&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
			
				<span class="k">if</span> <span class="n">input_search</span><span class="p">:</span>
					<span class="n">total_input_reads</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
	
				<span class="k">if</span> <span class="n">multi_search</span><span class="p">:</span>
					<span class="n">count_tmp</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
					<span class="n">count_multi</span> <span class="o">=</span> <span class="n">count_multi</span> <span class="o">+</span> <span class="n">count_tmp</span>
	
				<span class="k">elif</span> <span class="n">unmap_search</span><span class="p">:</span>
					<span class="n">perc_tmp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
					<span class="n">count_reads</span> <span class="o">=</span> <span class="n">math_functions</span><span class="o">.</span><span class="n">percentage</span><span class="p">(</span><span class="n">perc_tmp</span><span class="p">,</span> <span class="n">total_input_reads</span><span class="p">)</span>
					<span class="n">count_unmap</span> <span class="o">=</span> <span class="n">count_unmap</span> <span class="o">+</span> <span class="n">count_reads</span>
		<span class="k">else</span><span class="p">:</span>
	
			<span class="c1">## -------------------------------- ##</span>
			<span class="c1">## tophat</span>
			<span class="c1">## -------------------------------- ##</span>
	
			<span class="n">mapping_stats</span> <span class="o">=</span> <span class="n">mapping_folder</span> <span class="o">+</span> <span class="s1">&#39;/align_summary.txt&#39;</span> 
			<span class="n">count_map</span> <span class="o">=</span> <span class="mi">0</span>
			<span class="n">total_input_reads</span> <span class="o">=</span> <span class="mi">0</span>
			
			<span class="k">if</span> <span class="n">files_functions</span><span class="o">.</span><span class="n">is_non_zero_file</span><span class="p">(</span><span class="n">mapping_stats</span><span class="p">):</span>
				<span class="c1">## debugging messages</span>
				<span class="k">if</span> <span class="n">Debug</span><span class="p">:</span>
					<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;** DEBUG:&quot;</span><span class="p">)</span>
					<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;tophat mapping available for sample: &quot;</span> <span class="o">+</span> <span class="n">name</span><span class="p">)</span>
					<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;mapping_folder: &quot;</span> <span class="o">+</span> <span class="n">mapping_folder</span><span class="p">)</span>
				
				<span class="n">mapping_stats_file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">mapping_stats</span><span class="p">)</span>
				<span class="n">mapping_stats_file_text</span> <span class="o">=</span> <span class="n">mapping_stats_file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
				<span class="n">mapping_stats_file_lines</span> <span class="o">=</span> <span class="n">mapping_stats_file_text</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>	
	
				<span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">mapping_stats_file_lines</span><span class="p">:</span>
					<span class="n">map_search2</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;Aligned.*\:\s+(\d+).*&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
					<span class="n">input_search2</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;.*Input.*\:\s+(\d+).*&quot;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
					<span class="k">if</span> <span class="n">input_search2</span><span class="p">:</span>
						<span class="n">total_input_reads</span> <span class="o">=</span> <span class="n">input_search2</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
					<span class="k">if</span> <span class="n">map_search2</span><span class="p">:</span>
						<span class="n">count_map</span> <span class="o">=</span> <span class="n">map_search2</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
		
				<span class="c1">####</span>
				<span class="n">count_unmap</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">total_input_reads</span><span class="p">)</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">count_map</span><span class="p">)</span>
	
			<span class="k">else</span><span class="p">:</span>
				<span class="c1">## other</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;Neither tophat or STAR..., no mapping statistics&quot;</span><span class="p">)</span>
	
		<span class="c1">### print mapping stats</span>
		<span class="n">string2write_unmap</span> <span class="o">=</span> <span class="s2">&quot;unmapped</span><span class="se">\t</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span><span class="n">count_unmap</span>
		<span class="n">out_tsv_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">string2write_unmap</span><span class="p">)</span>
		
		<span class="c1">## close files</span>
		<span class="n">out_tsv_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

		<span class="c1">## print timestamp</span>
		<span class="n">time_functions</span><span class="o">.</span><span class="n">print_time_stamp</span><span class="p">(</span><span class="n">filename_stamp_parse</span><span class="p">)</span>

	<span class="k">return</span><span class="p">(</span><span class="n">out_tsv_file_name</span><span class="p">,</span> <span class="n">RNA_biotypes_file_name</span><span class="p">)</span></div>

<span class="c1">#######################################################################</span>
<div class="viewcode-block" id="RNAbiotype_module_call"><a class="viewcode-back" href="../../../devel/scripts/RNAbiotype.html#XICRA.scripts.RNAbiotype.RNAbiotype_module_call">[docs]</a><span class="k">def</span> <span class="nf">RNAbiotype_module_call</span><span class="p">(</span><span class="n">samples_dict</span><span class="p">,</span> <span class="n">output_dict</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">,</span> <span class="n">Debug</span><span class="p">,</span> <span class="n">max_workers_int</span><span class="p">,</span> <span class="n">threads_job</span><span class="p">,</span> <span class="n">multimapping</span><span class="p">,</span> <span class="n">stranded</span><span class="p">):</span>
	<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">	Create RNAbiotype analysis for each sample and create summary plots</span>
<span class="sd">	</span>
<span class="sd">	:param samples_dict: Dictionary containing sample IDs as keys and bam files as values</span>
<span class="sd">	:param output_dict: Dictionary containing sample IDs as keys and output folder as values</span>
<span class="sd">	:param gtf_file: Gene annotation file for the reference genome used.</span>
<span class="sd">	:param threads: Number of threads to use.</span>
<span class="sd">	:param Debug: True/False for debugging messages</span>
<span class="sd">	&quot;&quot;&quot;</span>
	
	<span class="c1">## get bin</span>
	<span class="n">featureCount_exe</span> <span class="o">=</span> <span class="n">set_config</span><span class="o">.</span><span class="n">get_exe</span><span class="p">(</span><span class="s1">&#39;featureCounts&#39;</span><span class="p">)</span>

	<span class="c1">## send for each sample</span>
	<span class="k">with</span> <span class="n">concurrent</span><span class="o">.</span><span class="n">futures</span><span class="o">.</span><span class="n">ThreadPoolExecutor</span><span class="p">(</span><span class="n">max_workers</span><span class="o">=</span><span class="n">max_workers_int</span><span class="p">)</span> <span class="k">as</span> <span class="n">executor</span><span class="p">:</span>
		<span class="n">commandsSent</span> <span class="o">=</span> <span class="p">{</span> <span class="n">executor</span><span class="o">.</span><span class="n">submit</span><span class="p">(</span><span class="n">biotype_all</span><span class="p">,</span> <span class="n">featureCount_exe</span><span class="p">,</span> 
										<span class="n">output_dict</span><span class="p">[</span><span class="n">sample</span><span class="p">],</span> <span class="n">gtf_file</span><span class="p">,</span> <span class="n">bam_files</span><span class="p">,</span> 
										<span class="n">sample</span><span class="p">,</span> <span class="n">threads_job</span><span class="p">,</span> <span class="n">Debug</span><span class="p">,</span> <span class="n">multimapping</span><span class="p">,</span> <span class="n">stranded</span><span class="p">):</span> <span class="n">sample</span> <span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">bam_files</span> <span class="ow">in</span> <span class="n">samples_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="p">}</span>
	
		<span class="k">for</span> <span class="n">cmd2</span> <span class="ow">in</span> <span class="n">concurrent</span><span class="o">.</span><span class="n">futures</span><span class="o">.</span><span class="n">as_completed</span><span class="p">(</span><span class="n">commandsSent</span><span class="p">):</span>
			<span class="n">details</span> <span class="o">=</span> <span class="n">commandsSent</span><span class="p">[</span><span class="n">cmd2</span><span class="p">]</span>
			<span class="k">try</span><span class="p">:</span>
				<span class="n">data</span> <span class="o">=</span> <span class="n">cmd2</span><span class="o">.</span><span class="n">result</span><span class="p">()</span>
			<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
				<span class="nb">print</span> <span class="p">(</span><span class="s1">&#39;***ERROR:&#39;</span><span class="p">)</span>
				<span class="nb">print</span> <span class="p">(</span><span class="n">cmd2</span><span class="p">)</span>
				<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">%r</span><span class="s1"> generated an exception: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">details</span><span class="p">,</span> <span class="n">exc</span><span class="p">))</span>

	<span class="c1">##</span>
	<span class="c1">## plot results</span>
	<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">folder</span> <span class="ow">in</span> <span class="n">output_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
		<span class="n">RNAbiotypes_stats_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="n">name</span> <span class="o">+</span> <span class="s1">&#39;_RNAbiotype.tsv&#39;</span><span class="p">)</span>
		<span class="k">if</span> <span class="n">files_functions</span><span class="o">.</span><span class="n">is_non_zero_file</span><span class="p">(</span><span class="n">RNAbiotypes_stats_file</span><span class="p">):</span>
			<span class="n">pie_plot_results</span><span class="p">(</span><span class="n">RNAbiotypes_stats_file</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">folder</span><span class="p">,</span> <span class="n">Debug</span><span class="p">)</span>
			
	<span class="k">return</span><span class="p">()</span></div>

<span class="c1">############################################################</span>
<div class="viewcode-block" id="generate_matrix"><a class="viewcode-back" href="../../../devel/scripts/RNAbiotype.html#XICRA.scripts.RNAbiotype.generate_matrix">[docs]</a><span class="k">def</span> <span class="nf">generate_matrix</span><span class="p">(</span><span class="n">dict_files</span><span class="p">):</span>
	<span class="c1">######################################################</span>
	<span class="c1">### For a dictionary containing names as keys and files as values, </span>
	<span class="c1">###	generates a count matrix with the classification from featurecounts</span>
	<span class="c1">###		</span>
	<span class="c1">### It returns a dataframe containing for each index generated</span>
	<span class="c1">### count values for each sample (sample_name) in columns</span>
	<span class="c1">###</span>
	<span class="c1">#########################################################</span>
	<span class="n">all_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
	<span class="k">for</span> <span class="n">key</span><span class="p">,</span><span class="n">values</span> <span class="ow">in</span> <span class="n">dict_files</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
		<span class="nb">print</span> <span class="p">(</span><span class="s1">&#39;+ Reading information from sample: &#39;</span><span class="p">,</span> <span class="n">key</span><span class="p">)</span>	
		<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">values</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;RNAbiotypes&#39;</span><span class="p">,</span> <span class="n">key</span><span class="p">])</span>
		
		<span class="c1">## skip if file is empty</span>
		<span class="k">if</span> <span class="n">data</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
			<span class="k">continue</span>

		<span class="c1">## get info, generate unique name and merge for samples</span>
		<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;RNAbiotypes&#39;</span><span class="p">)</span>
		<span class="n">all_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">all_data</span><span class="p">,</span> <span class="n">data</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	
	<span class="k">return</span> <span class="p">(</span><span class="n">all_data</span><span class="p">)</span></div>
	<span class="c1">##</span>

<span class="c1">#######################################################################</span>
<div class="viewcode-block" id="pie_plot_results"><a class="viewcode-back" href="../../../devel/scripts/RNAbiotype.html#XICRA.scripts.RNAbiotype.pie_plot_results">[docs]</a><span class="k">def</span> <span class="nf">pie_plot_results</span><span class="p">(</span><span class="n">RNAbiotypes_stats_file</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">folder</span><span class="p">,</span> <span class="n">Debug</span><span class="p">):</span>
	
	<span class="c1">##</span>
	<span class="n">filename_stamp_plot</span> <span class="o">=</span> <span class="n">folder</span> <span class="o">+</span> <span class="s1">&#39;/.success_plot&#39;</span>
	<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename_stamp_plot</span><span class="p">):</span>
		<span class="n">stamp</span> <span class="o">=</span> <span class="n">time_functions</span><span class="o">.</span><span class="n">read_time_stamp</span><span class="p">(</span><span class="n">filename_stamp_plot</span><span class="p">)</span>
		<span class="nb">print</span> <span class="p">(</span><span class="n">colored</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">A previous command generated results on: </span><span class="si">%s</span><span class="s2"> [</span><span class="si">%s</span><span class="s2"> -- </span><span class="si">%s</span><span class="s2">]&quot;</span> <span class="o">%</span><span class="p">(</span><span class="n">stamp</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="s1">&#39;plot results&#39;</span><span class="p">),</span> <span class="s1">&#39;yellow&#39;</span><span class="p">))</span>
	<span class="k">else</span><span class="p">:</span>
	
		<span class="c1"># PLOT and SHOW results</span>
		<span class="n">RNAbiotypes_stats</span> <span class="o">=</span> <span class="n">main_functions</span><span class="o">.</span><span class="n">get_data</span><span class="p">(</span><span class="n">RNAbiotypes_stats_file</span><span class="p">,</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="s1">&#39;header=None&#39;</span><span class="p">)</span>
	
		<span class="c1"># create plot</span>
		<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span>
		<span class="n">df_genetype_2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Type&#39;</span><span class="p">:</span><span class="n">RNAbiotypes_stats</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> 
									<span class="s1">&#39;Count&#39;</span><span class="p">:</span><span class="n">RNAbiotypes_stats</span><span class="p">[</span><span class="mi">1</span><span class="p">]})</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">])</span>
	
		<span class="c1">## get total count</span>
		<span class="n">df_genetype_ReadCount_sum</span> <span class="o">=</span> <span class="n">df_genetype_2</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
	
		<span class="c1">## filter 1% values</span>
		<span class="n">minimun</span> <span class="o">=</span> <span class="n">df_genetype_ReadCount_sum</span> <span class="o">*</span> <span class="mf">0.01</span>
		<span class="n">df_genetype_filter_greater</span> <span class="o">=</span> <span class="n">df_genetype_2</span><span class="p">[</span> <span class="n">df_genetype_2</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">minimun</span> <span class="p">]</span>
		<span class="n">df_genetype_filter_smaller</span> <span class="o">=</span> <span class="n">df_genetype_2</span><span class="p">[</span> <span class="n">df_genetype_2</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">minimun</span> <span class="p">]</span>
	
		<span class="c1">## create %values</span>
		<span class="n">df_genetype_2</span><span class="p">[</span><span class="s1">&#39;Percentage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">df_genetype_2</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">df_genetype_ReadCount_sum</span><span class="o">*</span><span class="mi">100</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
		
		<span class="c1">## merge and generate Other class</span>
		<span class="n">df_genetype_filter_smaller_sum</span> <span class="o">=</span> <span class="n">df_genetype_filter_smaller</span><span class="p">[</span><span class="s1">&#39;Count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="c1">## total filter smaller</span>
		<span class="n">df_genetype_filter_greater2</span> <span class="o">=</span> <span class="n">df_genetype_filter_greater</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
			<span class="s1">&#39;Count&#39;</span><span class="p">:</span><span class="n">df_genetype_filter_smaller_sum</span><span class="p">,</span> 
			<span class="s1">&#39;Type&#39;</span><span class="p">:</span><span class="s1">&#39;Other&#39;</span><span class="p">},</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
	
		<span class="c1">## Create Pie Plot</span>
		<span class="n">ax1</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">121</span><span class="p">,</span> <span class="n">aspect</span><span class="o">=</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
		<span class="n">df_genetype_filter_greater2</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span>
			<span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;Count&#39;</span><span class="p">,</span> 
			<span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> 
			<span class="n">autopct</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%1.2f%%</span><span class="s1">&#39;</span><span class="p">,</span> 
			<span class="n">shadow</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
			<span class="n">labels</span><span class="o">=</span><span class="n">df_genetype_filter_greater2</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">],</span> 
			<span class="n">legend</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
	
		<span class="c1"># plot table</span>
		<span class="n">ax2</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">122</span><span class="p">)</span>
		<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
		<span class="n">tbl</span> <span class="o">=</span> <span class="n">ax2</span><span class="o">.</span><span class="n">table</span><span class="p">(</span>
			<span class="n">cellText</span><span class="o">=</span><span class="n">df_genetype_2</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> 
			<span class="n">colLabels</span><span class="o">=</span><span class="n">df_genetype_2</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span>
			<span class="n">loc</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">rowLoc</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">cellLoc</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> 
			<span class="p">)</span>
		<span class="n">tbl</span><span class="o">.</span><span class="n">auto_set_font_size</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
		<span class="c1">#tbl.set_fontsize(12)</span>
		<span class="n">tbl</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="mf">1.1</span><span class="p">,</span><span class="mf">1.1</span><span class="p">)</span>
	
		<span class="c1">## set PDF name</span>
		<span class="n">name_figure</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="n">name</span> <span class="o">+</span> <span class="s1">&#39;_RNAbiotypes.pdf&#39;</span><span class="p">)</span>
	
		<span class="c1">## generate image</span>
		<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">name_figure</span><span class="p">)</span>
		<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="n">name_figure</span><span class="p">)</span>

		<span class="c1">## print time stamps</span>
		<span class="n">time_functions</span><span class="o">.</span><span class="n">print_time_stamp</span><span class="p">(</span><span class="n">filename_stamp_plot</span><span class="p">)</span>
		<span class="n">filename_stamp_all</span> <span class="o">=</span> <span class="n">folder</span> <span class="o">+</span> <span class="s1">&#39;/.success_all&#39;</span>
		<span class="n">time_functions</span><span class="o">.</span><span class="n">print_time_stamp</span><span class="p">(</span><span class="n">filename_stamp_all</span><span class="p">)</span></div>
		
<span class="c1">#######################################################################</span>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../devel/scripts/RNAbiotype.html#XICRA.scripts.RNAbiotype.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
	
	<span class="c1">## ARGV</span>
	<span class="k">if</span> <span class="nb">len</span> <span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">6</span><span class="p">:</span>
		<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Usage:&quot;</span><span class="p">)</span>
		<span class="nb">print</span> <span class="p">(</span><span class="s2">&quot;python3 </span><span class="si">%s</span><span class="s2"> bam_file folder gtf_file threads name featureCount_bin multimapping[True/False]</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">realpath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span>
		<span class="n">exit</span><span class="p">()</span>
	
	<span class="n">bam_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
	<span class="n">folder</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
	<span class="n">gtf_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">argv</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
	<span class="n">threads</span> <span class="o">=</span> <span class="n">argv</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
	<span class="n">name</span> <span class="o">=</span> <span class="n">argv</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
	<span class="n">featureCount_exe</span> <span class="o">=</span> <span class="n">argv</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
	<span class="n">multimapping</span><span class="o">=</span> <span class="n">argv</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>

	<span class="c1">## Debug</span>
	<span class="n">Debug</span><span class="o">=</span><span class="kc">True</span>
	
	<span class="c1">## variables</span>
	<span class="n">biotype_all</span><span class="p">(</span><span class="n">featureCount_exe</span><span class="p">,</span> <span class="n">folder</span><span class="p">,</span> <span class="n">gtf_file</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">threads</span><span class="p">,</span> <span class="n">Debug</span><span class="p">,</span> <span class="n">multimapping</span><span class="p">)</span>
	<span class="c1">## plot results</span>
	<span class="n">RNAbiotypes_stats_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">folder</span><span class="p">,</span> <span class="n">name</span> <span class="o">+</span> <span class="s1">&#39;_RNAbiotype.tsv&#39;</span><span class="p">)</span>
	<span class="k">if</span> <span class="n">files_functions</span><span class="o">.</span><span class="n">is_non_zero_file</span><span class="p">(</span><span class="n">RNAbiotypes_stats_file</span><span class="p">):</span>
		<span class="n">pie_plot_results</span><span class="p">(</span><span class="n">RNAbiotypes_stats_file</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">folder</span><span class="p">,</span> <span class="n">Debug</span><span class="p">)</span></div>
			
	
<span class="c1">######</span>
<span class="k">if</span> <span class="vm">__name__</span><span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
	<span class="n">main</span><span class="p">()</span>
	
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../../index.html">XICRA</a></h1>








<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../contents.html">Contents</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../index.html">Documentation overview</a><ul>
  <li><a href="../../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Jose F Sanchez-Herrero.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>